File: windows/UniviewCodeBuddyProxy/Services/LogBuffer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Channels;

namespace UniviewCodeBuddyProxy.Services;

/// <summary>Log level enum.</summary>
public enum LogLevel
{
    Debug,
    Info,
    Warn,
    Error,
}

/// <summary>Log entry record.</summary>
public sealed class LogEntry
{
    public Guid Id { get; } = Guid.NewGuid();
    public DateTime Timestamp { get; }
    public LogLevel Level { get; }
    public string Message { get; }

    public LogEntry(LogLevel level, string message, DateTime? timestamp = null)
    {
        Level = level;
        Message = message;
        Timestamp = timestamp ?? DateTime.UtcNow;
    }

    public string SseTimestamp => Timestamp.ToString("o");
    public string ShortTime => Timestamp.ToLocalTime().ToString("HH:mm:ss.fff");
}

/// <summary>
/// Thread-safe ring buffer log storage with SSE streaming and file logging.
/// </summary>
public sealed class LogBuffer : IDisposable
{
    // ═══ Constants ═══

    private const int Capacity = 10_000;
    private static readonly TimeSpan HeartbeatInterval = TimeSpan.FromSeconds(15);

    // ═══ Ring Buffer State ═══

    private readonly LogEntry?[] _ring = new LogEntry[Capacity];
    private int _writeIndex;
    private bool _hasWrapped;
    private readonly object _ringLock = new();

    // ═══ SSE Subscribers ═══

    private readonly Dictionary<Guid, Channel<LogEvent>> _subscribers = new();
    private readonly object _subLock = new();

    // ═══ File Logging ═══

    private readonly object _fileLock = new();
    private string? _logFilePath;
    private long _currentLogSize;
    private int _maxLogSizeMB = Constants.Defaults.LogMaxSizeMB;

    // ═══ Log Event for SSE ═══

    public enum LogEventType { Entry, Heartbeat }

    public readonly struct LogEvent
    {
        public LogEventType Type { get; }
        public LogEntry? Entry { get; }

        private LogEvent(LogEventType type, LogEntry? entry = null)
        {
            Type = type;
            Entry = entry;
        }

        public static LogEvent FromEntry(LogEntry entry) => new(LogEventType.Entry, entry);
        public static LogEvent Heartbeat() => new(LogEventType.Heartbeat);
    }

    // ═══ Constructor ═══

    public LogBuffer()
    {
        SetupLogFile();
        StartHeartbeatLoop();
    }

    // ═══ Append Methods ═══

    public void Append(LogLevel level, string message)
    {
        var entry = new LogEntry(level, message);
        AppendEntry(entry);
    }

    public void Info(string message) => Append(LogLevel.Info, message);
    public void Warn(string message) => Append(LogLevel.Warn, message);
    public void Error(string message) => Append(LogLevel.Error, message);
    public void Debug(string message) => Append(LogLevel.Debug, message);

    private void AppendEntry(LogEntry entry)
    {
        lock (_ringLock)
        {
            var idx = _writeIndex % Capacity;
            _ring[idx] = entry;
            _writeIndex++;
            if (_writeIndex >= Capacity && !_hasWrapped)
                _hasWrapped = true;
        }

        WriteToFile(entry);
        EmitToSubscribers(LogEvent.FromEntry(entry));
        EntryAppended?.Invoke(entry);
    }

    // ═══ Query Methods ═══

    public int Count
    {
        get { lock (_ringLock) { return _hasWrapped ? Capacity : _writeIndex; } }
    }

    public List<LogEntry> Recent(int count)
    {
        lock (_ringLock)
        {
            var total = _hasWrapped ? Capacity : _writeIndex;
            var start = Math.Max(0, total - count);
            var result = new List<LogEntry>(Math.Min(count, total));
            for (var i = start; i < total; i++)
            {
                var entry = _ring[i % Capacity];
                if (entry != null) result.Add(entry);
            }
            return result;
        }
    }

    public List<LogEntry> All() => Recent(Count);

    public List<LogEntry> Filter(LogLevel level)
    {
        lock (_ringLock)
        {
            var total = _hasWrapped ? Capacity : _writeIndex;
            var result = new List<LogEntry>();
            for (var i = 0; i < total; i++)
            {
                var entry = _ring[i % Capacity];
                if (entry != null && entry.Level == level)
                    result.Add(entry);
            }
            return result;
        }
    }

    public void Clear()
    {
        lock (_ringLock)
        {
            Array.Clear(_ring);
            _writeIndex = 0;
            _hasWrapped = false;
        }
    }

    /// <summary>Event fired when a new entry is appended.</summary>
    public event Action<LogEntry>? EntryAppended;

    // ═══ SSE Log Streaming ═══

    /// <summary>
    /// Subscribe to live log events. Returns a channel that receives backlog + live events + heartbeat.
    /// </summary>
    public Channel<LogEvent> Subscribe(int backlog = 100)
    {
        var channel = Channel.CreateUnbounded<LogEvent>();
        var subId = Guid.NewGuid();

        // Send backlog
        var history = Recent(backlog);
        foreach (var entry in history)
        {
            channel.Writer.TryWrite(LogEvent.FromEntry(entry));
        }

        lock (_subLock)
        {
            _subscribers[subId] = channel;
        }

        // Auto-remove on completion
        _ = Task.Run(async () =>
        {
            await channel.Reader.Completion;
            lock (_subLock)
            {
                _subscribers.Remove(subId);
            }
        });

        return channel;
    }

    /// <summary>
    /// Generate SSE format event text from a LogEvent.
    /// </summary>
    public string SseEventData(LogEvent logEvent)
    {
        switch (logEvent.Type)
        {
            case LogEventType.Entry when logEvent.Entry is { } entry:
                {
                    var escapedMsg = EscapeJson(entry.Message);
                    var data = $"{{\"timestamp\":\"{entry.SseTimestamp}\",\"level\":\"{entry.Level.ToString().ToLowerInvariant()}\",\"message\":{escapedMsg}}}";
                    return $"event: log\ndata: {data}\n\n";
                }
            case LogEventType.Heartbeat:
                {
                    var ts = DateTime.UtcNow.ToString("o");
                    return $"event: heartbeat\ndata: {{\"timestamp\":\"{ts}\"}}\n\n";
                }
            default:
                return "";
        }
    }

    private void EmitToSubscribers(LogEvent logEvent)
    {
        List<Channel<LogEvent>> channels;
        lock (_subLock)
        {
            channels = _subscribers.Values.ToList();
        }
        foreach (var ch in channels)
        {
            ch.Writer.TryWrite(logEvent);
        }
    }

    // ═══ Heartbeat ═══

    private Timer? _heartbeatTimer;

    private void StartHeartbeatLoop()
    {
        _heartbeatTimer = new Timer(_ =>
        {
            EmitToSubscribers(LogEvent.Heartbeat());
        }, null, HeartbeatInterval, HeartbeatInterval);
    }

    // ═══ File Logging ═══

    public int MaxLogSizeMB
    {
        get => _maxLogSizeMB;
        set => _maxLogSizeMB = value;
    }

    private void SetupLogFile()
    {
        try
        {
            var dir = Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
                "UniviewCodeBuddyProxy");
            Directory.CreateDirectory(dir);

            _logFilePath = Path.Combine(dir, "proxy.log");

            lock (_fileLock)
            {
                if (!File.Exists(_logFilePath))
                    File.Create(_logFilePath).Dispose();

                var fi = new FileInfo(_logFilePath);
                _currentLogSize = fi.Length;
            }
        }
        catch
        {
            // Ignore file setup errors
        }
    }

    private void WriteToFile(LogEntry entry)
    {
        if (_logFilePath == null) return;

        try
        {
            lock (_fileLock)
            {
                var line = $"{entry.ShortTime} [{entry.Level.ToString().ToUpperInvariant()}] {entry.Message}\n";
                var data = Encoding.UTF8.GetBytes(line);
                using var fs = new FileStream(_logFilePath, FileMode.Append, FileAccess.Write, FileShare.Read);
                fs.Write(data, 0, data.Length);
                _currentLogSize += data.Length;
                CheckAndTruncateIfNeeded();
            }
        }
        catch
        {
            // Ignore write errors
        }
    }

    private void CheckAndTruncateIfNeeded()
    {
        var maxBytes = (long)_maxLogSizeMB * 1024 * 1024;
        if (_currentLogSize <= maxBytes) return;

        try
        {
            // Truncate: clear file and reset
            using var fs = new FileStream(_logFilePath, FileMode.Create, FileAccess.Write, FileShare.Read);
            fs.Flush();
            _currentLogSize = 0;
        }
        catch
        {
            // Ignore truncation errors
        }
    }

    // ═══ JSON Escaping ═══

    private static string EscapeJson(string s)
    {
        var sb = new StringBuilder(s.Length + 8);
        sb.Append('"');
        foreach (var c in s)
        {
            switch (c)
            {
                case '\\': sb.Append("\\\\"); break;
                case '"': sb.Append("\\\""); break;
                case '\n': sb.Append("\\n"); break;
                case '\r': sb.Append("\\r"); break;
                case '\t': sb.Append("\\t"); break;
                default: sb.Append(c); break;
            }
        }
        sb.Append('"');
        return sb.ToString();
    }

    public void Dispose()
    {
        _heartbeatTimer?.Dispose();
    }
}

File: macos/UniviewCodeBuddyProxy/Sources/Services/LogBuffer.swift
import Foundation
import Combine

// ═══════════════════════════════════════════════
// LogEntry — 日志条目
// ═══════════════════════════════════════════════

struct LogEntry: Identifiable, Sendable {
    let id: UUID
    let timestamp: Date
    let level: LogLevel
    let message: String

    init(timestamp: Date = Date(), level: LogLevel, message: String) {
        self.id = UUID()
        self.timestamp = timestamp
        self.level = level
        self.message = message
    }

    /// SSE 格式的时间戳
    var sseTimestamp: String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter.string(from: timestamp)
    }

    /// 可读的简短时间
    var shortTime: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm:ss.SSS"
        return formatter.string(from: timestamp)
    }
}

// ═══════════════════════════════════════════════
// LogLevel — 日志级别
// ═══════════════════════════════════════════════

enum LogLevel: String, Sendable, CaseIterable {
    case debug = "debug"
    case info = "info"
    case warn = "warn"
    case error = "error"

    var emoji: String {
        switch self {
        case .debug: return "🔍"
        case .info:  return "ℹ️"
        case .warn:  return "⚠️"
        case .error: return "❌"
        }
    }

    /// 用于 SSE 事件的标签
    var sseTag: String {
        switch self {
        case .debug: return "debug"
        case .info:  return "info"
        case .warn:  return "warn"
        case .error: return "error"
        }
    }
}

// ═══════════════════════════════════════════════
// LogBuffer — 线程安全的环形日志缓冲区
// ═══════════════════════════════════════════════

final class LogBuffer: ObservableObject {

    // MARK: - 常量

    /// 环形缓冲区容量
    static let capacity = 10_000

    /// 心跳间隔（秒）
    private static let heartbeatInterval: TimeInterval = 15

    // MARK: - 线程安全存储

    /// 内部存储 actor，隔离并发访问
    private let storage = LogStorage()

    // MARK: - 订阅管理

    /// 活跃的 SSE 订阅者续命令牌
    private var subscriberContinuations: [UUID: AsyncStream<LogEvent>.Continuation] = [:]
    private let continuationsLock = NSLock()

    // MARK: - 文件日志

    private let fileQueue = DispatchQueue(label: "com.codebuddy.logbuffer.file", qos: .utility)
    private var logFileHandle: FileHandle?
    private var currentLogSize: Int64 = 0

    /// 日志文件最大字节数（由 ConfigManager 驱动，默认 50MB）
    var maxLogSizeMB: Int = Defaults.logMaxSizeMB {
        didSet { maxLogSizeBytes = Int64(maxLogSizeMB) * 1024 * 1024 }
    }

    private var maxLogSizeBytes: Int64 = Int64(Defaults.logMaxSizeMB) * 1024 * 1024

    // MARK: - @Published（SwiftUI 绑定）

    /// 最近的日志条目快照（供 SwiftUI 列表显示，最多保留 500 条）
    @Published private(set) var recentEntries: [LogEntry] = []

    /// 当前缓冲区中的总条目数
    @Published private(set) var totalEntryCount: Int = 0

    // MARK: - 日志事件（用于 SSE 流）

    enum LogEvent: Sendable {
        case entry(LogEntry)
        case heartbeat
    }

    // MARK: - Init

    init() {
        setupLogFile()
    }

    deinit {
        try? logFileHandle?.close()
    }

    // MARK: - 追加日志

    /// 追加一条日志
    func append(_ level: LogLevel, _ message: String) {
        let entry = LogEntry(level: level, message: message)
        appendEntry(entry)
    }

    /// 追加 info 级别日志
    func info(_ message: String) {
        append(.info, message)
    }

    /// 追加 warn 级别日志
    func warn(_ message: String) {
        append(.warn, message)
    }

    /// 追加 error 级别日志
    func error(_ message: String) {
        append(.error, message)
    }

    /// 追加 debug 级别日志
    func debug(_ message: String) {
        append(.debug, message)
    }

    private func appendEntry(_ entry: LogEntry) {
        Task {
            await storage.append(entry)

            // 预先在 actor 中计算，避免在 MainActor.run 闭包中 await
            let newCount = await storage.count
            let snapshot = await storage.recent(500)

            // 更新 @Published（主线程）
            await MainActor.run {
                self.totalEntryCount = newCount
                self.recentEntries = snapshot
            }

            // 写入文件
            writeToFile(entry)

            // 推送给 SSE 订阅者
            emitToSubscribers(.entry(entry))
        }
    }

    // MARK: - 查询

    /// 获取最近 N 条日志
    func recent(_ count: Int) async -> [LogEntry] {
        await storage.recent(count)
    }

    /// 获取所有缓冲区内的日志
    func allEntries() async -> [LogEntry] {
        await storage.all()
    }

    /// 按级别过滤
    func entries(matchingLevel level: LogLevel) async -> [LogEntry] {
        await storage.filter { $0.level == level }
    }

    // MARK: - SSE 日志流

    /// 创建一个 SSE 日志流：先发送历史积压，然后实时推送 + 心跳
    func logStream(backlog: Int = 100) -> AsyncStream<LogEvent> {
        AsyncStream { continuation in
            let subId = UUID()

            // 发送历史积压
            Task {
                let history = await storage.recent(backlog)
                for entry in history {
                    continuation.yield(.entry(entry))
                }
            }

            // 注册续命令牌
            continuationsLock.lock()
            subscriberContinuations[subId] = continuation
            continuationsLock.unlock()

            continuation.onTermination = { [weak self] _ in
                self?.continuationsLock.lock()
                self?.subscriberContinuations.removeValue(forKey: subId)
                self?.continuationsLock.unlock()
            }
        }
    }

    /// 生成 SSE 格式的事件文本
    func sseEventData(for event: LogEvent) -> String {
        switch event {
        case .entry(let entry):
            let data = """
            {"timestamp":"\(entry.sseTimestamp)","level":"\(entry.level.rawValue)","message":\(escapeJSON(entry.message))}
            """
            return "event: log\ndata: \(data)\n\n"

        case .heartbeat:
            return "event: heartbeat\ndata: {\"timestamp\":\"\(ISO8601DateFormatter().string(from: Date()))\"}\n\n"
        }
    }

    /// 启动心跳定时器（应在 SSE 连接建立后调用）
    func startHeartbeat() -> Task<Void, Never> {
        Task {
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: UInt64(Self.heartbeatInterval) * 1_000_000_000)
                guard !Task.isCancelled else { return }
                emitToSubscribers(.heartbeat)
            }
        }
    }

    // MARK: - 订阅者广播

    private func emitToSubscribers(_ event: LogEvent) {
        continuationsLock.lock()
        let continuations = subscriberContinuations.values
        continuationsLock.unlock()

        for continuation in continuations {
            continuation.yield(event)
        }
    }

    // MARK: - 文件日志

    private func setupLogFile() {
        fileQueue.async { [weak self] in
            self?.performLogFileSetup()
        }
    }

    private func performLogFileSetup() {
        let home = FileManager.default.homeDirectoryForCurrentUser.path
        let dir = "\(home)/.codebuddy-proxy"

        // 确保目录存在
        try? FileManager.default.createDirectory(atPath: dir, withIntermediateDirectories: true)

        let path = "\(dir)/proxy.log"

        // 如果文件不存在则创建
        if !FileManager.default.fileExists(atPath: path) {
            FileManager.default.createFile(atPath: path, contents: nil)
        }

        // 打开文件句柄
        if let handle = FileHandle(forWritingAtPath: path) {
            handle.seekToEndOfFile()
            self.currentLogSize = Int64(handle.offsetInFile)
            self.logFileHandle = handle
        }

        // 检查是否需要截断
        checkAndTruncateIfNeeded()
    }

    private func writeToFile(_ entry: LogEntry) {
        fileQueue.async { [weak self] in
            self?.performWrite(entry)
        }
    }

    private func performWrite(_ entry: LogEntry) {
        guard let handle = logFileHandle else { return }

        let line = "\(entry.shortTime) [\(entry.level.rawValue.uppercased())] \(entry.message)\n"
        guard let data = line.data(using: .utf8) else { return }

        handle.write(data)
        currentLogSize += Int64(data.count)

        checkAndTruncateIfNeeded()
    }

    private func checkAndTruncateIfNeeded() {
        guard currentLogSize > maxLogSizeBytes else { return }

        // 截断：关闭文件 → 清空 → 重新打开
        try? logFileHandle?.close()
        logFileHandle = nil

        let home = FileManager.default.homeDirectoryForCurrentUser.path
        let path = "\(home)/.codebuddy-proxy/proxy.log"

        // 清空文件内容
        FileManager.default.createFile(atPath: path, contents: nil)

        if let handle = FileHandle(forWritingAtPath: path) {
            self.logFileHandle = handle
            self.currentLogSize = 0
        }
    }

    // MARK: - 清空缓冲区

    /// 清空内存中的所有日志条目
    func clear() async {
        await storage.clear()
        await MainActor.run {
            self.recentEntries = []
            self.totalEntryCount = 0
        }
    }

    // MARK: - JSON 转义

    private func escapeJSON(_ string: String) -> String {
        var escaped = string
        escaped = escaped.replacingOccurrences(of: "\\", with: "\\\\")
        escaped = escaped.replacingOccurrences(of: "\"", with: "\\\"")
        escaped = escaped.replacingOccurrences(of: "\n", with: "\\n")
        escaped = escaped.replacingOccurrences(of: "\r", with: "\\r")
        escaped = escaped.replacingOccurrences(of: "\t", with: "\\t")
        return "\"\(escaped)\""
    }
}

// ═══════════════════════════════════════════════
// LogStorage — Actor 隔离的环形缓冲区内部存储
// ═══════════════════════════════════════════════

actor LogStorage {

    /// 环形缓冲区
    private var ring: [LogEntry] = []

    /// 写入位置（模运算）
    private var writeIndex: Int = 0

    /// 是否已写满一圈
    private var hasWrapped: Bool = false

    /// 容量
    private let capacity: Int

    init(capacity: Int = LogBuffer.capacity) {
        self.capacity = capacity
        self.ring = []
        ring.reserveCapacity(capacity)
    }

    /// 当前存储的条目数
    var count: Int {
        hasWrapped ? capacity : writeIndex
    }

    /// 追加一条日志
    func append(_ entry: LogEntry) {
        if writeIndex < capacity {
            ring.append(entry)
        } else {
            let idx = writeIndex % capacity
            ring[idx] = entry
        }
        writeIndex += 1
        if writeIndex >= capacity && !hasWrapped {
            hasWrapped = true
        }
    }

    /// 获取最近 N 条日志（按时间升序）
    func recent(_ count: Int) -> [LogEntry] {
        let total = self.count
        let start = max(0, total - count)

        var result: [LogEntry] = []
        result.reserveCapacity(min(count, total))

        for i in start..<total {
            let idx = i % capacity
            result.append(ring[idx])
        }
        return result
    }

    /// 获取所有日志（按时间升序）
    func all() -> [LogEntry] {
        recent(count)
    }

    /// 按条件过滤
    func filter(_ isIncluded: (LogEntry) -> Bool) -> [LogEntry] {
        var result: [LogEntry] = []
        let total = count
        for i in 0..<total {
            let idx = i % capacity
            let entry = ring[idx]
            if isIncluded(entry) {
                result.append(entry)
            }
        }
        return result
    }

    /// 清空
    func clear() {
        ring.removeAll(keepingCapacity: true)
        writeIndex = 0
        hasWrapped = false
    }
}
